<a href="https://colab.research.google.com/github/Hashwineey/Data-Science-Project-2_Sales-Data-Analysis/blob/main/P146219_Project2_Data_Cleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Data Cleaning**

## Import libraries and load csv files

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
# Load libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.patches as patches
import seaborn as sns

In [3]:
# Load datasets
calendar = pd.read_csv('/content/drive/MyDrive/STQD6014_DataScience/Project 2/AdventureWorks_Calendar.csv')
customer = pd.read_csv('/content/drive/MyDrive/STQD6014_DataScience/Project 2/AdventureWorks_Customers.csv', encoding='latin1')
#encoding = 'latin1' because the file is not in UTF-8 encoding
prod_cat = pd.read_csv('/content/drive/MyDrive/STQD6014_DataScience/Project 2/AdventureWorks_Product_Categories.csv')
prod_subcat = pd.read_csv('/content/drive/MyDrive/STQD6014_DataScience/Project 2/AdventureWorks_Product_Subcategories.csv')
prod = pd.read_csv('/content/drive/MyDrive/STQD6014_DataScience/Project 2/AdventureWorks_Products.csv')
returns = pd.read_csv('/content/drive/MyDrive/STQD6014_DataScience/Project 2/AdventureWorks_Returns.csv')
sales_2015 = pd.read_csv('/content/drive/MyDrive/STQD6014_DataScience/Project 2/AdventureWorks_Sales_2015.csv')
sales_2016 = pd.read_csv('/content/drive/MyDrive/STQD6014_DataScience/Project 2/AdventureWorks_Sales_2016.csv')
sales_2017 = pd.read_csv('/content/drive/MyDrive/STQD6014_DataScience/Project 2/AdventureWorks_Sales_2017.csv')
territories = pd.read_csv('/content/drive/MyDrive/STQD6014_DataScience/Project 2/AdventureWorks_Territories.csv')

In [4]:
# Print datasets' dimensions

# Dictionary to hold datasets
datasets = {
    "calendar": calendar,
    "customer": customer,
    "prod_cat": prod_cat,
    "prod_subcat": prod_subcat,
    "prod": prod,
    "returns": returns,
    "sales_2015": sales_2015,
    "sales_2016": sales_2016,
    "sales_2017": sales_2017,
    "territories": territories
}

# Loop through datasets and print shapes
for name, df in datasets.items():
    print(f"{name}: {df.shape}")

calendar: (876, 1)
customer: (18148, 13)
prod_cat: (4, 2)
prod_subcat: (37, 3)
prod: (293, 11)
returns: (1809, 4)
sales_2015: (2630, 8)
sales_2016: (23935, 8)
sales_2017: (29481, 8)
territories: (10, 4)


## Preview of each dataset

In [5]:
calendar.head()

Date
0  01-01-2015
1  01-02-2015
2  01-03-2015
3  01-04-2015
4  01-05-2015

In [6]:
customer.head()

CustomerKey Prefix  FirstName LastName   BirthDate MaritalStatus Gender  \
0        11000    MR.        JON     YANG  08-04-1966             M      M   
1        11001    MR.     EUGENE    HUANG  14-05-1965             S      M   
2        11002    MR.      RUBEN   TORRES  12-08-1965             M      M   
3        11003    MS.    CHRISTY      ZHU  15-02-1968             S      F   
4        11004   MRS.  ELIZABETH  JOHNSON  08-08-1968             S      F   

                     EmailAddress AnnualIncome  TotalChildren EducationLevel  \
0       jon24@adventure-works.com     $90,000               2      Bachelors   
1    eugene10@adventure-works.com     $60,000               3      Bachelors   
2     ruben35@adventure-works.com     $60,000               3      Bachelors   
3   christy12@adventure-works.com     $70,000               0      Bachelors   
4  elizabeth5@adventure-works.com     $80,000               5      Bachelors   

     Occupation HomeOwner  
0  Professional         Y  
1  Professional         N  
2  Professional         Y  
3  Professional         N  
4  Professional         Y

In [7]:
prod_cat.head()

ProductCategoryKey CategoryName
0                   1        Bikes
1                   2   Components
2                   3     Clothing
3                   4  Accessories

In [8]:
prod_subcat.head()

ProductSubcategoryKey  SubcategoryName  ProductCategoryKey
0                      1   Mountain Bikes                   1
1                      2       Road Bikes                   1
2                      3    Touring Bikes                   1
3                      4       Handlebars                   2
4                      5  Bottom Brackets                   2

In [9]:
prod.head()

ProductKey  ProductSubcategoryKey ProductSKU              ProductName  \
0         214                     31  HL-U509-R    Sport-100 Helmet, Red   
1         215                     31    HL-U509  Sport-100 Helmet, Black   
2         218                     23  SO-B909-M   Mountain Bike Socks, M   
3         219                     23  SO-B909-L   Mountain Bike Socks, L   
4         220                     31  HL-U509-B   Sport-100 Helmet, Blue   

             ModelName                                 ProductDescription  \
0            Sport-100  Universal fit, well-vented, lightweight , snap...   
1            Sport-100  Universal fit, well-vented, lightweight , snap...   
2  Mountain Bike Socks  Combination of natural and synthetic fibers st...   
3  Mountain Bike Socks  Combination of natural and synthetic fibers st...   
4            Sport-100  Universal fit, well-vented, lightweight , snap...   

  ProductColor ProductSize ProductStyle  ProductCost  ProductPrice  
0          Red           0            0      13.0863       34.9900  
1        Black           0            0      12.0278       33.6442  
2        White           M            U       3.3963        9.5000  
3        White           L            U       3.3963        9.5000  
4         Blue           0            0      12.0278       33.6442

In [10]:
returns.head()

ReturnDate  TerritoryKey  ProductKey  ReturnQuantity
0  18-01-2015             9         312               1
1  18-01-2015            10         310               1
2  21-01-2015             8         346               1
3  22-01-2015             4         311               1
4  02-02-2015             6         312               1

In [11]:
sales_2015.head()

OrderDate   StockDate OrderNumber  ProductKey  CustomerKey  TerritoryKey  \
0  01-01-2015  21-09-2001     SO45080         332        14657             1   
1  01-01-2015  05-12-2001     SO45079         312        29255             4   
2  01-01-2015  29-10-2001     SO45082         350        11455             9   
3  01-01-2015  16-11-2001     SO45081         338        26782             6   
4  02-01-2015  15-12-2001     SO45083         312        14947            10   

   OrderLineItem  OrderQuantity  
0              1              1  
1              1              1  
2              1              1  
3              1              1  
4              1              1

In [12]:
sales_2016.head()

OrderDate   StockDate OrderNumber  ProductKey  CustomerKey  TerritoryKey  \
0  01-01-2016  17-10-2002     SO48797         385        14335             1   
1  01-01-2016  30-09-2002     SO48802         383        24923             9   
2  01-01-2016  29-11-2002     SO48801         326        15493             1   
3  01-01-2016  16-11-2002     SO48799         352        26708             4   
4  01-01-2016  16-12-2002     SO48798         369        23332             9   

   OrderLineItem  OrderQuantity  
0              1              1  
1              1              1  
2              1              1  
3              1              1  
4              1              1

In [13]:
sales_2017.head()

OrderDate   StockDate OrderNumber  ProductKey  CustomerKey  TerritoryKey  \
0  01-01-2017  13-12-2003     SO61285         529        23791             1   
1  01-01-2017  24-09-2003     SO61285         214        23791             1   
2  01-01-2017  04-09-2003     SO61285         540        23791             1   
3  01-01-2017  28-09-2003     SO61301         529        16747             1   
4  01-01-2017  21-10-2003     SO61301         377        16747             1   

   OrderLineItem  OrderQuantity  
0              2              2  
1              3              1  
2              1              1  
3              2              2  
4              1              1

In [14]:
territories.head()

SalesTerritoryKey     Region        Country      Continent
0                  1  Northwest  United States  North America
1                  2  Northeast  United States  North America
2                  3    Central  United States  North America
3                  4  Southwest  United States  North America
4                  5  Southeast  United States  North America

## Data Preprosessing

### Merge all datasets

#### 1. Join product_category and product_subcategory dataframes on ProductCategoryKey.
##### Essentially, we want the ProductCategoryKey from product_category to map on product_subcategory. The shape of the df should be (37,4)

In [15]:
# Merge on 'ProductCategoryKey'
merged_df1 = pd.merge(prod_cat, prod_subcat, on='ProductCategoryKey', how='inner')

In [16]:
merged_df1.head()

ProductCategoryKey CategoryName  ProductSubcategoryKey  SubcategoryName
0                   1        Bikes                      1   Mountain Bikes
1                   1        Bikes                      2       Road Bikes
2                   1        Bikes                      3    Touring Bikes
3                   2   Components                      4       Handlebars
4                   2   Components                      5  Bottom Brackets

In [17]:
merged_df1.shape

(37, 4)

#### 2. Merge merged_df1 with product df on ProductSubcategoryKey. The shape of the df should be (293,14) since we are adding 3 columns from merged_df1 to the product dataframe.

In [18]:
# Merge on 'ProductSubcategoryKey'
merged_df2 = pd.merge(prod, merged_df1, on='ProductSubcategoryKey', how='inner')
merged_df2.shape

(293, 14)

In [19]:
merged_df2.head()

ProductKey  ProductSubcategoryKey ProductSKU              ProductName  \
0         214                     31  HL-U509-R    Sport-100 Helmet, Red   
1         215                     31    HL-U509  Sport-100 Helmet, Black   
2         218                     23  SO-B909-M   Mountain Bike Socks, M   
3         219                     23  SO-B909-L   Mountain Bike Socks, L   
4         220                     31  HL-U509-B   Sport-100 Helmet, Blue   

             ModelName                                 ProductDescription  \
0            Sport-100  Universal fit, well-vented, lightweight , snap...   
1            Sport-100  Universal fit, well-vented, lightweight , snap...   
2  Mountain Bike Socks  Combination of natural and synthetic fibers st...   
3  Mountain Bike Socks  Combination of natural and synthetic fibers st...   
4            Sport-100  Universal fit, well-vented, lightweight , snap...   

  ProductColor ProductSize ProductStyle  ProductCost  ProductPrice  \
0          Red           0            0      13.0863       34.9900   
1        Black           0            0      12.0278       33.6442   
2        White           M            U       3.3963        9.5000   
3        White           L            U       3.3963        9.5000   
4         Blue           0            0      12.0278       33.6442   

   ProductCategoryKey CategoryName SubcategoryName  
0                   4  Accessories         Helmets  
1                   4  Accessories         Helmets  
2                   3     Clothing           Socks  
3                   3     Clothing           Socks  
4                   4  Accessories         Helmets

#### 3. Merge merged_df2 with returns on ProductKey. The shape should be (1809,17) since we are adding 3 columns from returns to merged_df2

In [20]:
# Merge on 'ProductSubcategoryKey'
merged_df3 = pd.merge(merged_df2, returns, on='ProductKey', how='inner')
merged_df3.shape

(1809, 17)

#### 4. Add 'Sales_2015', 'Sales_2016' and 'Sales_2017' as separate columns in Customer dataset, based on the 'CustomerKey'

In [21]:
# Add a Year column to each sales dataset
sales_2015['Year'] = 2015
sales_2016['Year'] = 2016
sales_2017['Year'] = 2017

In [22]:
# Merge customer dataset with sales_2015 dataset. The shape should be (2630, 21)
cust_2015 = pd.merge(customer, sales_2015, on='CustomerKey', how='inner')

In [23]:
cust_2015.shape

(2630, 21)

In [24]:
# Merge customer dataset with sales_2016 dataset. The shape should be (23935, 21)
cust_2016 = pd.merge(customer, sales_2016, on='CustomerKey', how='inner')
cust_2016.shape

(23935, 21)

In [25]:
# Merge customer dataset with sales_2017 dataset. The shape should be (29481, 21)
cust_2017 = pd.merge(customer, sales_2017, on='CustomerKey', how='inner')
cust_2017.shape

(29481, 21)

In [26]:
# Merge datasets on 'CustomerKey'
merged_df = cust_2015.merge(cust_2016, how='outer')
merged_df4 = merged_df.merge(cust_2017, how='outer')
merged_df4.shape

(56046, 21)

In [27]:
merged_df4.head(3)

CustomerKey Prefix FirstName LastName   BirthDate MaritalStatus Gender  \
0        11000    MR.       JON     YANG  08-04-1966             M      M   
1        11000    MR.       JON     YANG  08-04-1966             M      M   
2        11000    MR.       JON     YANG  08-04-1966             M      M   

                EmailAddress AnnualIncome  TotalChildren  ...    Occupation  \
0  jon24@adventure-works.com     $90,000               2  ...  Professional   
1  jon24@adventure-works.com     $90,000               2  ...  Professional   
2  jon24@adventure-works.com     $90,000               2  ...  Professional   

  HomeOwner   OrderDate   StockDate OrderNumber ProductKey  TerritoryKey  \
0         Y  04-11-2016  06-09-2003     SO57418        488             9   
1         Y  04-11-2016  10-08-2003     SO57418        541             9   
2         Y  04-11-2016  11-08-2003     SO57418        530             9   

   OrderLineItem  OrderQuantity  Year  
0              5              1  2016  
1              2              2  2016  
2              3              2  2016  

[3 rows x 21 columns]

#### 5. Merge merged_df4 and merged_df3 on the common columns

In [28]:
# Get the common columns between merged_df4 and merged_df3
common_columns = set(merged_df4.columns).intersection(merged_df3.columns)

# Print the common columns
print(common_columns)

{'ProductKey', 'TerritoryKey'}


In [29]:
# Perform the merge based on the common columns
merged_df5 = pd.merge(merged_df4, merged_df3, on=list(common_columns), how='inner')
merged_df5.head()

CustomerKey Prefix FirstName LastName   BirthDate MaritalStatus Gender  \
0        11000    MR.       JON     YANG  08-04-1966             M      M   
1        11000    MR.       JON     YANG  08-04-1966             M      M   
2        11000    MR.       JON     YANG  08-04-1966             M      M   
3        11000    MR.       JON     YANG  08-04-1966             M      M   
4        11000    MR.       JON     YANG  08-04-1966             M      M   

                EmailAddress AnnualIncome  TotalChildren  ... ProductColor  \
0  jon24@adventure-works.com     $90,000               2  ...       Yellow   
1  jon24@adventure-works.com     $90,000               2  ...       Yellow   
2  jon24@adventure-works.com     $90,000               2  ...       Yellow   
3  jon24@adventure-works.com     $90,000               2  ...       Yellow   
4  jon24@adventure-works.com     $90,000               2  ...       Yellow   

  ProductSize ProductStyle ProductCost ProductPrice ProductCategoryKey  \
0           S            U     41.5723        53.99                  3   
1           S            U     41.5723        53.99                  3   
2           S            U     41.5723        53.99                  3   
3           S            U     41.5723        53.99                  3   
4           S            U     41.5723        53.99                  3   

   CategoryName  SubcategoryName  ReturnDate  ReturnQuantity  
0      Clothing          Jerseys  26-09-2016               1  
1      Clothing          Jerseys  06-10-2016               1  
2      Clothing          Jerseys  03-12-2016               1  
3      Clothing          Jerseys  15-01-2017               1  
4      Clothing          Jerseys  08-04-2017               1  

[5 rows x 36 columns]

In [30]:
merged_df5.shape

(424467, 36)

In [31]:
# To verify that there is no duplicates
duplicates = merged_df5[merged_df5.duplicated()]

# Display the duplicates
print(duplicates)

Empty DataFrame
Columns: [CustomerKey, Prefix, FirstName, LastName, BirthDate, MaritalStatus, Gender, EmailAddress, AnnualIncome, TotalChildren, EducationLevel, Occupation, HomeOwner, OrderDate, StockDate, OrderNumber, ProductKey, TerritoryKey, OrderLineItem, OrderQuantity, Year, ProductSubcategoryKey, ProductSKU, ProductName, ModelName, ProductDescription, ProductColor, ProductSize, ProductStyle, ProductCost, ProductPrice, ProductCategoryKey, CategoryName, SubcategoryName, ReturnDate, ReturnQuantity]
Index: []

[0 rows x 36 columns]


No duplicated rows

#### 6. Merge merged_df5 and territories dataset on TerritoryKey, the shape should be (424467, 39)

In [32]:
# Rename SalesTerritoryKey to TerritoryKey in territories dataset
territories = territories.rename(columns={'SalesTerritoryKey': 'TerritoryKey'})
territories.head()

TerritoryKey     Region        Country      Continent
0             1  Northwest  United States  North America
1             2  Northeast  United States  North America
2             3    Central  United States  North America
3             4  Southwest  United States  North America
4             5  Southeast  United States  North America

In [33]:
df = pd.merge(merged_df5, territories, on='TerritoryKey', how='inner')
df.head(3)

CustomerKey Prefix FirstName LastName   BirthDate MaritalStatus Gender  \
0        11000    MR.       JON     YANG  08-04-1966             M      M   
1        11000    MR.       JON     YANG  08-04-1966             M      M   
2        11000    MR.       JON     YANG  08-04-1966             M      M   

                EmailAddress AnnualIncome  TotalChildren  ... ProductCost  \
0  jon24@adventure-works.com     $90,000               2  ...     41.5723   
1  jon24@adventure-works.com     $90,000               2  ...     41.5723   
2  jon24@adventure-works.com     $90,000               2  ...     41.5723   

  ProductPrice ProductCategoryKey CategoryName SubcategoryName  ReturnDate  \
0        53.99                  3     Clothing         Jerseys  26-09-2016   
1        53.99                  3     Clothing         Jerseys  06-10-2016   
2        53.99                  3     Clothing         Jerseys  03-12-2016   

   ReturnQuantity     Region    Country  Continent  
0               1  Australia  Australia    Pacific  
1               1  Australia  Australia    Pacific  
2               1  Australia  Australia    Pacific  

[3 rows x 39 columns]

In [34]:
df.shape

(424467, 39)

### Data Cleaning

Now that we had merged all of the datasets (excluded calendar df since its not neccesary), we first need to identify the types of the data we have

In [35]:
pd.set_option('display.max_columns', None)
df.head(3)

CustomerKey Prefix FirstName LastName   BirthDate MaritalStatus Gender  \
0        11000    MR.       JON     YANG  08-04-1966             M      M   
1        11000    MR.       JON     YANG  08-04-1966             M      M   
2        11000    MR.       JON     YANG  08-04-1966             M      M   

                EmailAddress AnnualIncome  TotalChildren EducationLevel  \
0  jon24@adventure-works.com     $90,000               2      Bachelors   
1  jon24@adventure-works.com     $90,000               2      Bachelors   
2  jon24@adventure-works.com     $90,000               2      Bachelors   

     Occupation HomeOwner   OrderDate   StockDate OrderNumber  ProductKey  \
0  Professional         Y  04-11-2016  06-09-2003     SO57418         488   
1  Professional         Y  04-11-2016  06-09-2003     SO57418         488   
2  Professional         Y  04-11-2016  06-09-2003     SO57418         488   

   TerritoryKey  OrderLineItem  OrderQuantity  Year  ProductSubcategoryKey  \
0             9              5              1  2016                     21   
1             9              5              1  2016                     21   
2             9              5              1  2016                     21   

  ProductSKU                     ProductName                    ModelName  \
0  SJ-0194-S  Short-Sleeve Classic Jersey, S  Short-Sleeve Classic Jersey   
1  SJ-0194-S  Short-Sleeve Classic Jersey, S  Short-Sleeve Classic Jersey   
2  SJ-0194-S  Short-Sleeve Classic Jersey, S  Short-Sleeve Classic Jersey   

                                  ProductDescription ProductColor ProductSize  \
0  Short sleeve classic breathable jersey with su...       Yellow           S   
1  Short sleeve classic breathable jersey with su...       Yellow           S   
2  Short sleeve classic breathable jersey with su...       Yellow           S   

  ProductStyle  ProductCost  ProductPrice  ProductCategoryKey CategoryName  \
0            U      41.5723         53.99                   3     Clothing   
1            U      41.5723         53.99                   3     Clothing   
2            U      41.5723         53.99                   3     Clothing   

  SubcategoryName  ReturnDate  ReturnQuantity     Region    Country Continent  
0         Jerseys  26-09-2016               1  Australia  Australia   Pacific  
1         Jerseys  06-10-2016               1  Australia  Australia   Pacific  
2         Jerseys  03-12-2016               1  Australia  Australia   Pacific

In [42]:
# Identify the type of the data
print(df.dtypes)

CustomerKey                int64
Prefix                    object
FirstName                 object
LastName                  object
BirthDate                 object
MaritalStatus             object
Gender                    object
EmailAddress              object
AnnualIncome              object
TotalChildren              int64
EducationLevel            object
Occupation                object
HomeOwner                 object
OrderDate                 object
StockDate                 object
OrderNumber               object
ProductKey                 int64
TerritoryKey               int64
OrderLineItem              int64
OrderQuantity              int64
Year                       int64
ProductSubcategoryKey      int64
ProductSKU                object
ProductName               object
ModelName                 object
ProductDescription        object
ProductColor              object
ProductSize               object
ProductStyle              object
ProductCost              float64
ProductPri

In [43]:
df['CategoryName'].unique()

array(['Clothing', 'Accessories', 'Bikes'], dtype=object)

In [44]:
df['SubcategoryName'].unique()

array(['Jerseys', 'Tires and Tubes', 'Helmets', 'Touring Bikes',
       'Fenders', 'Mountain Bikes', 'Road Bikes', 'Bottles and Cages',
       'Caps', 'Bike Stands', 'Gloves', 'Shorts', 'Socks', 'Bike Racks',
       'Hydration Packs', 'Cleaners', 'Vests'], dtype=object)

In [45]:
df['ProductStyle'].unique()

array(['U', '0', 'W'], dtype=object)

#### 1. Manage null values (if any)

In [46]:
df.isna().sum()

CustomerKey                   0
Prefix                     2837
FirstName                     0
LastName                      0
BirthDate                     0
MaritalStatus                 0
Gender                     2837
EmailAddress                  0
AnnualIncome                  0
TotalChildren                 0
EducationLevel                0
Occupation                    0
HomeOwner                     0
OrderDate                     0
StockDate                     0
OrderNumber                   0
ProductKey                    0
TerritoryKey                  0
OrderLineItem                 0
OrderQuantity                 0
Year                          0
ProductSubcategoryKey         0
ProductSKU                    0
ProductName                   0
ModelName                     0
ProductDescription            0
ProductColor             312041
ProductSize                   0
ProductStyle                  0
ProductCost                   0
ProductPrice                  0
ProductCategoryKey            0
CategoryName                  0
SubcategoryName               0
ReturnDate                    0
ReturnQuantity                0
Region                        0
Country                       0
Continent                     0
dtype: int64

##### Handle 'Gender' null values

In [56]:
# Display all null values of Prefix
gender_null = df[df['Prefix'].isna()]
gender_null.head(3)

CustomerKey Prefix  FirstName LastName   BirthDate MaritalStatus Gender  \
1208        11025    NaN  ALEJANDRO     BECK  23-12-1945             M    NaN   
1209        11025    NaN  ALEJANDRO     BECK  23-12-1945             M    NaN   
1210        11025    NaN  ALEJANDRO     BECK  23-12-1945             M    NaN   

                         EmailAddress AnnualIncome  TotalChildren  \
1208  alejandro45@adventure-works.com     $10,000               2   
1209  alejandro45@adventure-works.com     $10,000               2   
1210  alejandro45@adventure-works.com     $10,000               2   

           EducationLevel Occupation HomeOwner   OrderDate   StockDate  \
1208  Partial High School   Clerical         Y  24-07-2016  09-04-2003   
1209  Partial High School   Clerical         Y  24-07-2016  09-04-2003   
1210  Partial High School   Clerical         Y  24-07-2016  09-04-2003   

     OrderNumber  ProductKey  TerritoryKey  OrderLineItem  OrderQuantity  \
1208     SO51561         485             9              2              2   
1209     SO51561         485             9              2              2   
1210     SO51561         485             9              2              2   

      Year  ProductSubcategoryKey ProductSKU            ProductName  \
1208  2016                     30    FE-6654  Fender Set - Mountain   
1209  2016                     30    FE-6654  Fender Set - Mountain   
1210  2016                     30    FE-6654  Fender Set - Mountain   

                  ModelName                        ProductDescription  \
1208  Fender Set - Mountain  Clip-on fenders fit most mountain bikes.   
1209  Fender Set - Mountain  Clip-on fenders fit most mountain bikes.   
1210  Fender Set - Mountain  Clip-on fenders fit most mountain bikes.   

     ProductColor ProductSize ProductStyle  ProductCost  ProductPrice  \
1208          NaN           0            0       8.2205         21.98   
1209          NaN           0            0       8.2205         21.98   
1210          NaN           0            0       8.2205         21.98   

      ProductCategoryKey CategoryName SubcategoryName  ReturnDate  \
1208                   4  Accessories         Fenders  13-07-2016   
1209                   4  Accessories         Fenders  05-09-2016   
1210                   4  Accessories         Fenders  17-10-2016   

      ReturnQuantity     Region    Country Continent  
1208               1  Australia  Australia   Pacific  
1209               1  Australia  Australia   Pacific  
1210               1  Australia  Australia   Pacific

In [57]:
len(gender_null['FirstName'].unique())

113

Total of 113 customers don't have their gender record. However, we can compute them based on their First Names.

In [58]:
gender_null['FirstName'].unique()

array(['ALEJANDRO', 'WENDY', 'ANGELA', 'ANNA', 'CESAR', 'KATHERINE',
       'NAOMI', 'KATIE', 'ALISHA', 'TODD', 'ADRIANA', 'DARREN', 'JASMINE',
       'ALEXA', 'ROBYN', 'BRANDON', 'RENEE', 'TERESA', 'CATHERINE',
       'NICOLAS', 'BENJAMIN', 'NINA', 'MINDY', 'ABBY', 'MICAH', 'RICARDO',
       'HAROLD', 'MIGUEL', 'ARTHUR', 'LEVI', 'ANDREA', 'RUTH', 'SHAWN',
       'NATHANIEL', 'AMANDA', 'NOAH', 'CEDRIC', 'MAKAYLA', 'LATOYA',
       'KEITH', 'MISTY', 'SHEILA', 'ALEXANDRIA', 'MADISON', 'DOUGLAS',
       'RICKY', 'JOSE', 'CHRISTIAN', 'RACHEL', 'JONATHAN', 'EBONY',
       'GLORIA', 'CANDICE', 'SHANNON', 'TAMMY', 'ORLANDO', 'FERNANDO',
       'JENNY', 'JACK', 'JASON', 'JÉSUS', 'LUIS', 'GERALD', 'EMILY',
       'SHEENA', 'DAISY', 'JACQUELINE', 'EMMA', 'SAVANNAH', 'CHRISTOPHER',
       'KRISTI', 'JEROME', 'MARIE', 'BARBARA', 'MARCUS', 'PEDRO',
       'CYNTHIA', 'DENNIS', 'JULIA', 'VANESSA', 'GAIL', 'ALEXANDRA',
       'THOMAS', 'EDDIE', 'BRETT', 'LESLIE', 'ZACHARY', 'CASEY', 'KENDRA',
       '

From the names above, we can list out all the male names.

In [59]:
male = ['DARREN','ARTHUR','CHRISTOPHER','LUIS','GERALD','JÉSUS','NICOLAS','CHRISTIAN','TODD','JOSE','MIGUEL',
       'JACK','NATHANIEL','DOUGLAS','DEVIN','LOUIS','KEITH','BRANDON','FERNANDO','MICAH','CESAR','GAIL','REGINALD',
       'CRAIG','ZACHARY','DALTON','ALEJANDRO','JAMIE','LESLIE','RICARDO','SHAWN','BENJAMIN','LEVI','TERRY','JONATHAN',
       'MARCUS','HAROLD','IAN','JEROME','RANDALL','ORLANDO','EDWARD','ALEXANDER','THOMAS','DENNIS','RICKY','PEDRO',
       'EDDIE','BRETT','WYATT','BARRY','NOAH','ROBYN','JASON','NEIL','CEDRIC']

# Fill null values of gender based on the names listed

df.loc[df['Gender'].isnull(), 'Gender'] = df.loc[df['Gender'].isnull(), 'FirstName'].map(
    lambda x: 'M' if x in male else 'F'
    )

df.isnull().sum()

CustomerKey                   0
Prefix                     2837
FirstName                     0
LastName                      0
BirthDate                     0
MaritalStatus                 0
Gender                        0
EmailAddress                  0
AnnualIncome                  0
TotalChildren                 0
EducationLevel                0
Occupation                    0
HomeOwner                     0
OrderDate                     0
StockDate                     0
OrderNumber                   0
ProductKey                    0
TerritoryKey                  0
OrderLineItem                 0
OrderQuantity                 0
Year                          0
ProductSubcategoryKey         0
ProductSKU                    0
ProductName                   0
ModelName                     0
ProductDescription            0
ProductColor             312041
ProductSize                   0
ProductStyle                  0
ProductCost                   0
ProductPrice                  0
ProductCategoryKey            0
CategoryName                  0
SubcategoryName               0
ReturnDate                    0
ReturnQuantity                0
Region                        0
Country                       0
Continent                     0
dtype: int64

In [60]:
# Check if the values are correctly replaced in Gender column
df[df['Prefix'].isnull()]

CustomerKey Prefix  FirstName LastName   BirthDate MaritalStatus  \
1208          11025    NaN  ALEJANDRO     BECK  23-12-1945             M   
1209          11025    NaN  ALEJANDRO     BECK  23-12-1945             M   
1210          11025    NaN  ALEJANDRO     BECK  23-12-1945             M   
1211          11025    NaN  ALEJANDRO     BECK  23-12-1945             M   
1212          11025    NaN  ALEJANDRO     BECK  23-12-1945             M   
...             ...    ...        ...      ...         ...           ...   
422846        29345    NaN    RANDALL     DIAZ  26-02-1942             S   
422847        29345    NaN    RANDALL     DIAZ  26-02-1942             S   
422848        29345    NaN    RANDALL     DIAZ  26-02-1942             S   
422849        29345    NaN    RANDALL     DIAZ  26-02-1942             S   
422850        29345    NaN    RANDALL     DIAZ  26-02-1942             S   

       Gender                     EmailAddress AnnualIncome  TotalChildren  \
1208        M  alejandro45@adventure-works.com     $10,000               2   
1209        M  alejandro45@adventure-works.com     $10,000               2   
1210        M  alejandro45@adventure-works.com     $10,000               2   
1211        M  alejandro45@adventure-works.com     $10,000               2   
1212        M  alejandro45@adventure-works.com     $10,000               2   
...       ...                              ...          ...            ...   
422846      M     randall4@adventure-works.com     $40,000               2   
422847      M     randall4@adventure-works.com     $40,000               2   
422848      M     randall4@adventure-works.com     $40,000               2   
422849      M     randall4@adventure-works.com     $40,000               2   
422850      M     randall4@adventure-works.com     $40,000               2   

             EducationLevel  Occupation HomeOwner   OrderDate   StockDate  \
1208    Partial High School    Clerical         Y  24-07-2016  09-04-2003   
1209    Partial High School    Clerical         Y  24-07-2016  09-04-2003   
1210    Partial High School    Clerical         Y  24-07-2016  09-04-2003   
1211    Partial High School    Clerical         Y  24-07-2016  09-04-2003   
1212    Partial High School    Clerical         Y  24-07-2016  09-04-2003   
...                     ...         ...       ...         ...         ...   
422846            Bachelors  Management         N  29-06-2017  03-04-2004   
422847            Bachelors  Management         N  29-06-2017  03-04-2004   
422848            Bachelors  Management         N  29-06-2017  03-04-2004   
422849            Bachelors  Management         N  29-06-2017  03-04-2004   
422850            Bachelors  Management         N  29-06-2017  03-04-2004   

       OrderNumber  ProductKey  TerritoryKey  OrderLineItem  OrderQuantity  \
1208       SO51561         485             9              2              2   
1209       SO51561         485             9              2              2   
1210       SO51561         485             9              2              2   
1211       SO51561         485             9              2              2   
1212       SO51561         485             9              2              2   
...            ...         ...           ...            ...            ...   
422846     SO74094         215             9              2              1   
422847     SO74094         215             9              2              1   
422848     SO74094         215             9              2              1   
422849     SO74094         215             9              2              1   
422850     SO74094         215             9              2              1   

        Year  ProductSubcategoryKey ProductSKU              ProductName  \
1208    2016                     30    FE-6654    Fender Set - Mountain   
1209    2016                     30    FE-6654    Fender Set - Mountain   
1210    2016                     30    FE-6654    Fender Set - Mountain

We can ignore the null values in 'ProductColor' and 'Prefix' since the columns are not part of the analysis

#### 2. Drop Unused Columns

In [61]:
# Drop unneccesary columns
# We want to exclude ['Prefix','FirstName','LastName','EmailAddress','OrderNumber','ProductKey',
#'OrderLineItem','ProductSubcategoryKey','ProductSKU','ProductDescription','ProductCategoryKey', 'ProductColor']

df = df.drop(['Prefix','FirstName','LastName','EmailAddress','OrderNumber','ProductKey',
              'OrderLineItem','ProductSubcategoryKey','ProductSKU', 'ProductDescription','ProductCategoryKey','ProductColor'],
             axis=1)
df.shape

(424467, 27)

#### 3. Get customers' age from their birth date. And then, bin the age into groups

In [62]:
df['BirthDate'].head()

0    08-04-1966
1    08-04-1966
2    08-04-1966
3    08-04-1966
4    08-04-1966
Name: BirthDate, dtype: object

In [63]:
from datetime import datetime

# Convert BirthDate to datetime
df['BirthDate'] = pd.to_datetime(df['BirthDate'], format='%d-%m-%Y', errors='coerce')

# Check the converted column
print(df['BirthDate'].head())

0   1966-04-08
1   1966-04-08
2   1966-04-08
3   1966-04-08
4   1966-04-08
Name: BirthDate, dtype: datetime64[ns]


In [64]:
# Define today's date in 'now' variable
now = datetime.now()
now

datetime.datetime(2025, 1, 26, 11, 42, 6, 791924)

In [65]:
# Calucate customers' age from 'BirthDate'
df['Age'] = (now.year - df['BirthDate'].dt.year) - ((now.month - df['BirthDate'].dt.month) < 0)
df['Age'].head()

0    58
1    58
2    58
3    58
4    58
Name: Age, dtype: int32

In [66]:
# We can drop BirthDate column now
df = df.drop(columns=['BirthDate'])
df.head(3)

CustomerKey MaritalStatus Gender AnnualIncome  TotalChildren  \
0        11000             M      M     $90,000               2   
1        11000             M      M     $90,000               2   
2        11000             M      M     $90,000               2   

  EducationLevel    Occupation HomeOwner   OrderDate   StockDate  \
0      Bachelors  Professional         Y  04-11-2016  06-09-2003   
1      Bachelors  Professional         Y  04-11-2016  06-09-2003   
2      Bachelors  Professional         Y  04-11-2016  06-09-2003   

   TerritoryKey  OrderQuantity  Year                     ProductName  \
0             9              1  2016  Short-Sleeve Classic Jersey, S   
1             9              1  2016  Short-Sleeve Classic Jersey, S   
2             9              1  2016  Short-Sleeve Classic Jersey, S   

                     ModelName ProductSize ProductStyle  ProductCost  \
0  Short-Sleeve Classic Jersey           S            U      41.5723   
1  Short-Sleeve Classic Jersey           S            U      41.5723   
2  Short-Sleeve Classic Jersey           S            U      41.5723   

   ProductPrice CategoryName SubcategoryName  ReturnDate  ReturnQuantity  \
0         53.99     Clothing         Jerseys  26-09-2016               1   
1         53.99     Clothing         Jerseys  06-10-2016               1   
2         53.99     Clothing         Jerseys  03-12-2016               1   

      Region    Country Continent  Age  
0  Australia  Australia   Pacific   58  
1  Australia  Australia   Pacific   58  
2  Australia  Australia   Pacific   58

In [75]:
age = np.sort(df['Age'].unique())
print(age)

[ 44  45  46  47  48  49  50  51  52  53  54  55  56  57  58  59  60  61
  62  63  64  65  66  67  68  69  70  71  72  73  74  75  76  77  78  79
  80  81  82  83  84  85  86  87  88  89  90  91  92  93  94  95  96  97
  98  99 100 101 102 103 104 105 106 107 108 109 111 112 113 114]


In [76]:
# Since we have customers whose age is more than 80, we will drop those customers as they are not our customer target group
# We need to know how many of them first
pd.set_option('display.max_rows', None)
count = df[df['Age'] > 80].shape[0]
print(count)

33650


In [77]:
# Drop rows where Age > 80
df = df[df['Age'] <= 80]
df.shape

(390817, 28)

In [78]:
# Get min and max of age
print("Max of Age: ", df['Age'].max())
print("Min of Age: ", df['Age'].min())
print("Median of Age: ", df['Age'].median())

Max of Age:  80
Min of Age:  44
Median of Age:  60.0


In [79]:
# Define the age class based on age
bins = [40,50,60,70,80]
df['AgeClass'] = pd.cut(df['Age'], bins=bins)

In [80]:
df['AgeClass'].value_counts()

AgeClass
(50, 60]    140360
(60, 70]    118739
(70, 80]     69765
(40, 50]     61953
Name: count, dtype: int64

#### 4. Convert all the date columns to datetime format

In [67]:
# 'OrderDate'
df['OrderDate'] = pd.to_datetime(df['OrderDate'], format='%d-%m-%Y', errors='coerce')

# 'StockDate'
df['StockDate'] = pd.to_datetime(df['StockDate'], format='%d-%m-%Y', errors='coerce')

# 'Year'
df['Year'] = pd.to_datetime(df['Year'], format='%Y', errors='coerce')

# 'ReturnDate'
df['ReturnDate'] = pd.to_datetime(df['ReturnDate'], format='%d-%m-%Y', errors='coerce')

In [68]:
# Check the format
print('OrderDate: ', df['OrderDate'].dtypes)
print('StockDate: ', df['StockDate'].dtypes)
print('Year: ', df['Year'].dtypes)
print('ReturnDate: ', df['ReturnDate'].dtypes)

OrderDate:  datetime64[ns]
StockDate:  datetime64[ns]
Year:  datetime64[ns]
ReturnDate:  datetime64[ns]


#### 5. Bin Annual Income

In [69]:
df['AnnualIncome'].dtypes

dtype('O')

In [70]:
# Remove dollar sign and commas, then convert to integer
df['AnnualIncome'] = df['AnnualIncome'].replace('[\$,]', '', regex=True).astype(int)

In [71]:
df['AnnualIncome'].head()

0    90000
1    90000
2    90000
3    90000
4    90000
Name: AnnualIncome, dtype: int64

In [72]:
# Get min and max of AnnualIncome
print("Max of AnnualIncome: ", df['AnnualIncome'].max())
print("Min of AnnualIncome: ", df['AnnualIncome'].min())
print("Median of AnnualIncome: ", df['AnnualIncome'].median())

Max of AnnualIncome:  170000
Min of AnnualIncome:  10000
Median of AnnualIncome:  60000.0


In [73]:
# Define the annual income class
bins = [0, 10000, 40000, 80000, 120000, 160000, 200000]
df['AnnualIncomeClass'] = pd.cut(df['AnnualIncome'], bins=bins)
df.head()

CustomerKey MaritalStatus Gender  AnnualIncome  TotalChildren  \
0        11000             M      M         90000              2   
1        11000             M      M         90000              2   
2        11000             M      M         90000              2   
3        11000             M      M         90000              2   
4        11000             M      M         90000              2   

  EducationLevel    Occupation HomeOwner  OrderDate  StockDate  TerritoryKey  \
0      Bachelors  Professional         Y 2016-11-04 2003-09-06             9   
1      Bachelors  Professional         Y 2016-11-04 2003-09-06             9   
2      Bachelors  Professional         Y 2016-11-04 2003-09-06             9   
3      Bachelors  Professional         Y 2016-11-04 2003-09-06             9   
4      Bachelors  Professional         Y 2016-11-04 2003-09-06             9   

   OrderQuantity       Year                     ProductName  \
0              1 2016-01-01  Short-Sleeve Classic Jersey, S   
1              1 2016-01-01  Short-Sleeve Classic Jersey, S   
2              1 2016-01-01  Short-Sleeve Classic Jersey, S   
3              1 2016-01-01  Short-Sleeve Classic Jersey, S   
4              1 2016-01-01  Short-Sleeve Classic Jersey, S   

                     ModelName ProductSize ProductStyle  ProductCost  \
0  Short-Sleeve Classic Jersey           S            U      41.5723   
1  Short-Sleeve Classic Jersey           S            U      41.5723   
2  Short-Sleeve Classic Jersey           S            U      41.5723   
3  Short-Sleeve Classic Jersey           S            U      41.5723   
4  Short-Sleeve Classic Jersey           S            U      41.5723   

   ProductPrice CategoryName SubcategoryName ReturnDate  ReturnQuantity  \
0         53.99     Clothing         Jerseys 2016-09-26               1   
1         53.99     Clothing         Jerseys 2016-10-06               1   
2         53.99     Clothing         Jerseys 2016-12-03               1   
3         53.99     Clothing         Jerseys 2017-01-15               1   
4         53.99     Clothing         Jerseys 2017-04-08               1   

      Region    Country Continent  Age AnnualIncomeClass  
0  Australia  Australia   Pacific   58   (80000, 120000]  
1  Australia  Australia   Pacific   58   (80000, 120000]  
2  Australia  Australia   Pacific   58   (80000, 120000]  
3  Australia  Australia   Pacific   58   (80000, 120000]  
4  Australia  Australia   Pacific   58   (80000, 120000]

In [74]:
df['AnnualIncomeClass'].value_counts()

AnnualIncomeClass
(40000, 80000]      195599
(10000, 40000]      130301
(80000, 120000]      58213
(0, 10000]           19582
(120000, 160000]     17951
(160000, 200000]      2821
Name: count, dtype: int64

#### 6. Calculate Product Profit from ProductCost and ProductPrice

In [83]:
df['ProductProfit'] = (df['ProductPrice'] - df['ProductCost']).round(2)
df.head(3)

CustomerKey MaritalStatus Gender  AnnualIncome  TotalChildren  \
0        11000             M      M         90000              2   
1        11000             M      M         90000              2   
2        11000             M      M         90000              2   

  EducationLevel    Occupation HomeOwner  OrderDate  StockDate  TerritoryKey  \
0      Bachelors  Professional         Y 2016-11-04 2003-09-06             9   
1      Bachelors  Professional         Y 2016-11-04 2003-09-06             9   
2      Bachelors  Professional         Y 2016-11-04 2003-09-06             9   

   OrderQuantity       Year                     ProductName  \
0              1 2016-01-01  Short-Sleeve Classic Jersey, S   
1              1 2016-01-01  Short-Sleeve Classic Jersey, S   
2              1 2016-01-01  Short-Sleeve Classic Jersey, S   

                     ModelName ProductSize ProductStyle  ProductCost  \
0  Short-Sleeve Classic Jersey           S            U      41.5723   
1  Short-Sleeve Classic Jersey           S            U      41.5723   
2  Short-Sleeve Classic Jersey           S            U      41.5723   

   ProductPrice CategoryName SubcategoryName ReturnDate  ReturnQuantity  \
0         53.99     Clothing         Jerseys 2016-09-26               1   
1         53.99     Clothing         Jerseys 2016-10-06               1   
2         53.99     Clothing         Jerseys 2016-12-03               1   

      Region    Country Continent  Age AnnualIncomeClass  AgeClass  \
0  Australia  Australia   Pacific   58   (80000, 120000]  (50, 60]   
1  Australia  Australia   Pacific   58   (80000, 120000]  (50, 60]   
2  Australia  Australia   Pacific   58   (80000, 120000]  (50, 60]   

   ProductProfit UsagePeriod  
0          12.42    -39 days  
1          12.42    -29 days  
2          12.42     29 days

#### 7. Convert OrderDate from YYYY-MM-DD to YYYY-MM

In [84]:
# Format OrderDate as YYYY-MM using strftime
df['OrderDate'] = df['OrderDate'].dt.strftime('%Y-%m')

# Display the updated DataFrame
print(df[['OrderDate']].head())

  OrderDate
0   2016-11
1   2016-11
2   2016-11
3   2016-11
4   2016-11


#### 8. Again, drop unnecessary columns

In [86]:
df = df.drop(['StockDate','TerritoryKey','ReturnDate', 'TotalChildren','HomeOwner','ProductName','ProductSize','ProductStyle','Age'], axis=1)
df.head(3)

CustomerKey MaritalStatus Gender  AnnualIncome EducationLevel  \
0        11000             M      M         90000      Bachelors   
1        11000             M      M         90000      Bachelors   
2        11000             M      M         90000      Bachelors   

     Occupation OrderDate  OrderQuantity       Year  \
0  Professional   2016-11              1 2016-01-01   
1  Professional   2016-11              1 2016-01-01   
2  Professional   2016-11              1 2016-01-01   

                     ModelName  ProductCost  ProductPrice CategoryName  \
0  Short-Sleeve Classic Jersey      41.5723         53.99     Clothing   
1  Short-Sleeve Classic Jersey      41.5723         53.99     Clothing   
2  Short-Sleeve Classic Jersey      41.5723         53.99     Clothing   

  SubcategoryName  ReturnQuantity     Region    Country Continent  \
0         Jerseys               1  Australia  Australia   Pacific   
1         Jerseys               1  Australia  Australia   Pacific   
2         Jerseys               1  Australia  Australia   Pacific   

  AnnualIncomeClass  AgeClass  ProductProfit UsagePeriod  
0   (80000, 120000]  (50, 60]          12.42    -39 days  
1   (80000, 120000]  (50, 60]          12.42    -29 days  
2   (80000, 120000]  (50, 60]          12.42     29 days

In [87]:
df.isnull().sum()

CustomerKey          0
MaritalStatus        0
Gender               0
AnnualIncome         0
EducationLevel       0
Occupation           0
OrderDate            0
OrderQuantity        0
Year                 0
ModelName            0
ProductCost          0
ProductPrice         0
CategoryName         0
SubcategoryName      0
ReturnQuantity       0
Region               0
Country              0
Continent            0
AnnualIncomeClass    0
AgeClass             0
ProductProfit        0
UsagePeriod          0
dtype: int64

#### 9. Drop duplicated rows

In [88]:
# Find and drop duplicated rows

duplicates = df[df.duplicated()]
duplicates.shape

(85, 22)

In [89]:
# Drop duplicates in place
df.drop_duplicates(inplace=True)
df.shape

(390732, 22)

### Export cleaned dataset for data analysis

In [90]:
# Save dataframe to CSV in Google Drive
path = '/content/drive/MyDrive/STQD6014_DataScience/Project 2/cleaned_data.csv'
df.to_csv(path, index=False)